docs/DesignDocs/ExtendedEnclaveInitializationData.md

```diff
@@ -0,0 +1,154 @@
+
+
+Motivations
+----------
+
+With current SGX quotes, applications can attest their configuration as part of the `reportData` argument of `oe_get_report`.
+This mechanism works robustly when the application code is entirely and statically known to the validator of the quote.
+
+Things get more complicated if an enclave executes user code that is not part of the signed enclave image. For instance, consider an enclave containing a JavaScript interpreter that executes user scripts from the host and has access to the `oe_get_report` API. It is impossible to know which script has been executed by such an enclave based on a standard quote, even if the hash of the script is included in the `reportData`, because a malicious script can obtain a valid quote with `oe_get_report` pretending to be a honest script. Similarly, if an enclave loads and execute arbitrary assembly code from the host, this assembly code can use the `EGETQUOTE` instruction to create valid reports - hence, it is impossible to use attestation to tell which assembly code has been loaded by the enclave.
+
+An important instance of this problem is Azure Confidential Containers: the base enclave image contains the SGX-LKL runtime, which executes a user container. The container contains arbitrary code, which can use the `EGETQUOTE` instruction to obtain a valid quote for arbitrary `reportData` – potentially impersonating other containers.
+
+One way to solve this issue is to re-compile and re-sign a new enclave image for every container we execute (with the container data being measured together with the LKL base). However, this approach is cumbersome as we would like to be able to launch arbitrary containers on-demand without having to build and sign a new enclave image every time.
+
+Another class of problems we address is dynamic attestation of the memory and thread configuration. In current OE attestation, these settings (`NumStackPage`, `NumHeapPages`, `NumTCS`) are part of the configuration to sign an enclave. This means that it is not possible to deploy the same enclave image with different memory and core configurations.
+
+User Experience
+---------------
+
+The user decides to enable the use of EEID when the enclave base image is signed: we require that EEID enclaves use `NumStackPages=0`, `NumHeapPages=0`, and `NumTCS=0`.
+This guarantees that enclave images meant to be used with EEID cannot be accidentally initialized with normal attestation (the enclave will fail to load).
+
+To launch an EEID enclave, the user needs to use the new `oe_create_enclave_eeid` API:
+```C
+oe_result_t oe_create_enclave_eeid(
+    const char* enclave_path,
+    oe_enclave_type_t enclave_type,
+    uint32_t flags,
+    const oe_enclave_setting_t* settings,
+    uint32_t setting_count,
+    const oe_ocall_func_t* ocall_table,
+    uint32_t ocall_count,
+    oe_eeid_t* eeid,
```

Comment:
I would instead recommend a design that makes layered attestation be a first class concept, and is integrated with the attestation plugin design.

Reply:
I'm working on plugins at the moment, I think we can do most of what we want to do inside the plugin infrastructure, except for the bit that injects the additional attested data pages.

Reply:
It looks like we'll go for an experimental API for now, but yes, we can also put the EEID into the settings.

Reply:
I hope you understand the hassle of introducing a new OE public API and then later retracting it. Generally we'd try to avoid that.

Reply:
Yes, I do, and that's why I proposed new functions called `*_eeid` instead of modifying existing ones. Moving the extra argument into the settings of the existing functions effectively modifies the API too.

Reply:
Sure, sounds good. I've changed the document accordingly.

Reply:
Yes, the customizable loader has been proposed before, that's a good way to treat it, I think.

Reply:
Despite the semantic overloading of `oe_enclave_setting_t` between pre-load and post-load configuration, I am OK with putting EEID there to minimize the impact on the public API.